### ISSUE TITLE
Feature Request: Obey MaxDepth in SerializeObject()

### ISSUE THREAD

**Issue by b9chris** [JamesNK/Newtonsoft.Json#2191]
It would be helpful if MaxDepth worked on SerializeObject() as well, not just Deserialize().

Currently, you can use MaxDepth to stop the serializer from deserializing anything past a given depth on the initial object(s). But, you can't currently tell it to only serialize down to a certain depth.

Additionally, it appears that the normal behavior of MaxDepth is to throw an exception, which is not desirable. It would help if there was a way to tell it to simply serialize/deserialize down to MaxDepth and stop. An exception is undesirable behavior in many circumstances.

A good example of where this is useful is on an Entity Framework database with LazyLoading enabled and virtual properties throughout. `SerializeObject(user, new JsonSerializerSettings { MaxDepth = 0 });` would be useful for containing JSON serialization to just the user and their FK Ids. Without MaxDepth, JsonConvert is going to dutifully serialize most or all of the entire database.

I'm happy to write this feature if people are interested in it.

**Comment by myesn** [comment]
Hi, does json.net support this feature now?

**Comment by jdlundquist** [comment]
I'd be interested in `MaxDepth` setting (on `JsonSerializerSettings` and `JsonSerializer`)  working for `JsonConvert.SerializeObject()`. Ideally it would work for all overloads (The initial overload I'm specifically interested in using it for is: `JsonConvert.SerializeObject(object?, JsonSerializerSettings)`.

A starting point would be to have the serialization behavior mirror existing deserialization behavior (i.e. throwing exception if that's correct per @b9chris observation).

A second level of enhancement could have both serialization and serialization that respects `MaxDepth` with options on what to do when MaxDepth is exceeded. It could be similar to [ReferenceLoopHandling setting](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_ReferenceLoopHandling.htm),but in this case with just `Error` (default/existing) or `Ignore` options.

**Comment by alikrc** [comment]
I've same problem, when serializing at 2nd depth i get error. Any workaround to solve this problem? 

Found an old solution but not sure if can be done in another way
https://stackoverflow.com/questions/10453127/how-to-serialize-or-deserialize-a-json-object-to-a-certain-depth-in-c

**Comment by Mike-Logit** [comment]
I really need this too for logging purposes. My log messages are huge :/

**Comment by PaulDMendoza** [comment]
How is this not fixed? Everyone who has set this expects a certain behavior and isn't getting it.

**Comment by 5FEKO** [comment]
@b9chris we (The World) are very interested! Please go ahead! Thanks for your contribution.

**Comment by Premier-Solutions** [comment]
> I really need this too for logging purposes. My log messages are huge :/

did you find any work around for this issue. I'm logging all updates (CUD) and navigation properties are also getting serialized making the log size huge.

**Comment by Mike-Logit** [comment]
> did you find any work around for this issue. 

Yes but it's not pretty. I had to create my own JsonConverter and pass it to `JsonConvert.SerializeObject`.

For my use case, I wanted to log the contents of an object that contained data for executing a command or query, which I called the "Request" object:

```c-sharp
var rawRequest = JsonConvert.SerializeObject(request, Formatting.None, new RequestJsonConverter(2));
```

And this is my `RequestJsonConverter` class:

```c-sharp
/// <summary>
/// Converts the request object without any complex class property values included
/// </summary>
private class RequestJsonConverter : JsonConverter
{
    private readonly int _depth;

    public RequestJsonConverter(int depth)
    {
        _depth = depth;
    }

    public override bool CanRead => false;

    public override bool CanConvert(Type objectType) => true;

    public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
    {
        var jObject = new JObject();
        var requestProperties = value.GetType().GetProperties();
        RequestJsonConverter nextConverter = null;

        foreach (var prop in requestProperties)
        {
            if (_depth <= 0)
            {
                continue;
            }

            try
            {
                var propValue = prop.GetValue(value);
                if (!prop.PropertyType.IsPrimitive &&
                    !prop.PropertyType.IsValueType &&
                    prop.PropertyType != typeof(string))
                {
                    if (propValue is null || _depth <= 1)
                    {
                        continue;
                    }

                    nextConverter ??= new RequestJsonConverter(_depth - 1);
                    var serializedPropValue = JsonConvert.SerializeObject(propValue, Formatting.None, nextConverter);
                    jObject.Add(prop.Name, serializedPropValue);
                }
                else
                {
                    jObject.Add(prop.Name, propValue.ToSafeString());
                }
            }
            catch
            {
                // ignored - property getter might have thrown an exception so skip it.
            }
        }

        jObject.WriteTo(writer);
    }

    public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
        => throw new NotImplementedException("Unsupported");
}
```
